src/transformers/models/gpt2/modeling_flax_gpt2.py

```diff
@@ -0,0 +1,645 @@
+            max_length (:obj:`int`):
+                maximum possible length for auto-regressive decoding. Defines the sequence length of the initialized
+                cache.
+        """
+        # init input variables to retrieve cache
+        input_ids = jnp.ones((batch_size, max_length))
+        attention_mask = jnp.ones_like(input_ids)
+        position_ids = jnp.broadcast_to(jnp.arange(jnp.atleast_2d(input_ids).shape[-1]), input_ids.shape)
+
+        init_variables = self.module.init(
+            jax.random.PRNGKey(0), input_ids, attention_mask, position_ids, return_dict=False, _use_cache=True
+        )
+        return init_variables["cache"]
+
+    @add_start_docstrings_to_model_forward(GPT2_INPUTS_DOCSTRING)
+    def __call__(
+        self,
+        input_ids,
+        attention_mask=None,
+        position_ids=None,
+        params: dict = None,
+        past_key_values: dict = None,
+        dropout_rng: jax.random.PRNGKey = None,
+        train: bool = False,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+    ):
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.return_dict
+
+        batch_size, sequence_length = input_ids.shape
+
+        if position_ids is None:
+            if past_key_values is not None and input_ids.shape[-1] == 1:
+                # if `past_key_values` are passed and input_ids are
+                # longer than 1, we are in cached auto-regressive
+                # generation. It has to be made sure that position_ids
+                # are set correctly
+                cache_shift = next(
+                    iter(v for k, v in flatten_dict(unfreeze(past_key_values)).items() if k[-1] == "cache_index")
+                )
+                position_ids = jnp.broadcast_to(
+                    jnp.arange(self.config.max_position_embeddings)[None, :],
+                    (batch_size, self.config.max_position_embeddings),
+                )
+                position_ids = lax.dynamic_slice(position_ids, (0, cache_shift), (batch_size, 1))
+            else:
+                position_ids = jnp.broadcast_to(jnp.arange(sequence_length)[None, :], (batch_size, sequence_length))
+
+        if attention_mask is None:
+            # if past_key_values are passed we need to create an
+            # attention_mask of the same length as `cache_length`
+            if past_key_values is not None:
+                cache_length = next(
+                    iter(v for k, v in flatten_dict(unfreeze(past_key_values)).items() if k[-1] != "cache_index")
+                ).shape[1]
```

Comment:
This is really a weird way of getting the cache length.

Reply:
Good point - I could also just take the first one tensor in the cache => that will be more readable